### PR TITLE
Adds Neon Carpet

### PR DESCRIPTION
**Pull request by TemporalOroboros**
## About The Pull Request
- Adds a couple varieties of neon carpet.
- Makes decals care about their plane in addition to their layer.
- Probably doesn't 150% maptick this time.

<details>
<summary>Images (Click me)</summary>

Dayshift:
neon-carpets-dayshift

Nightshift:
neon-carpets-nightshift

Power Outage:
neon-carpets-poweroutage

Carpet colors and edge lighting are subject to change.
</details>

<details>
<summary>Sources:</summary>

All variants of simple neon carpets can be made with carpet, tinea luxor, and a dye:
- White: Made with sodium
- Red: Made with mindbreaker toxin
- Orange: Made with vitrium froth
- Yellow: Made with stabilizing agent
- Lime: Made with lime juice
- Green: Made with unstable mutagen
- Cyan: Made with salbutamol
- Teal: Made with nicotine
- Blue: Made with happiness
- Purple: Made with hyper-plasmium oxide
- Violet: Made with helbital
- Pink: Made with impedrezine
- Black: Made with multiver

Alternatively, all simple neon carpet varieties are available at cargo for a small sum.
The arcade cabinet can also dispense one variety of emissive floor as a prize.

</details>

<details>
<summary>Issues:</summary>

- [x] ~~The emissive blockers for airlocks straight up aren't working.~~ (Fixed)
- [X] ~~The emissive blockers for closets aren't including the door.~~ (Fixed)
- [X] ~~The carpet tile inhand emissive overlays aren't working.~~ (Fixed)
- [X] ~~The new human emissive blockers are blocking out the preferences character previews.~~ (Fixed)

</details>

## Why It's Good For The Game
- More aesthetic variety for people to decorate the station with.
- Gives coders the ability to define decal planes.

## Changelog
:cl:
add: Neon carpets exists. Bug your local chemist or quartermaster for a supply.
code: Decals support user-defined planes and some forms of smoothing.
/:cl:

### CODE REVIEW

**Comment by TemporalOroboros** [comment]
Please do not merge this until it has been TMed and profiled.

**Comment by Kylerace** [comment]
this time its probably fine since emissive blockers are overlays now

File: code/__DEFINES/icon_smoothing.dm [hunk 1]
```diff
@@ -64,14 +64,29 @@ DEFINE_BITFIELD(smoothing_flags, list(
...
+
+#define SMOOTH_GROUP_CLOSED_TURFS S_TURF(38) ///turf/closed
+#define SMOOTH_GROUP_MATERIAL_WALLS S_TURF(39) ///turf/closed/wall/material
+#define SMOOTH_GROUP_SYNDICATE_WALLS S_TURF(40) ///turf/closed/wall/r_wall/syndicate, /turf/closed/indestructible/syndicate
+#define SMOOTH_GROUP_HOTEL_WALLS S_TURF(41) ///turf/closed/indestructible/hotelwall
+#define SMOOTH_GROUP_MINERAL_WALLS S_TURF(42) ///turf/closed/mineral, /turf/closed/indestructible
+#define SMOOTH_GROUP_BOSS_WALLS S_TURF(43) ///turf/closed/indestructible/riveted/boss
+#define SMOOTH_GROUP_SURVIVAL_TITANIUM_WALLS S_TURF(44) ///turf/closed/wall/mineral/titanium/survival
 
 #define MAX_S_TURF SMOOTH_GROUP_SURVIVAL_TITANIUM_WALLS //Always match this value with the one above it.
```
Comment (Mothblocks) [inline code/__DEFINES/icon_smoothing.dm:91]: Always match this value with the one above it.

Comment (TemporalOroboros) [reply code/__DEFINES/icon_smoothing.dm:91]: Isn't it already matched?

Comment (Mothblocks) [reply code/__DEFINES/icon_smoothing.dm:91]: Didn't notice you put the new ones in the middle and not the end, my bad.

**Comment by TemporalOroboros** [comment]
I have no idea how a static variable is being used before it is initialized but it's breaking the tests.

**Comment by TemporalOroboros** [comment]
I have patched out the runtime. Apparently `/static` variables get initialized after `GLOB` for some reason and this breaks things if you try to use them.

**Comment by GoldenAlpharex** [comment]
I'm curious, could there be a way to disable the pattern it makes in the middle when there's other carpet on all sides? Either by making a different option when using it in-hand like the other floortiles, or screwdriving it when it's in your hands, or something like that, because at a small scale, it might look okay, but on a large scale it just doesn't look as sharp anymore.

**Comment by TemporalOroboros** [comment]
@GoldenAlpharex I could do that pretty easily, but it would double the number of sprites.

**Comment by GoldenAlpharex** [comment]
> @GoldenAlpharex I could do that pretty easily, but it would double the number of sprites.

Wouldn't it just require a neutral black one without any edges for the centers, and just have all the carpets be able to use that same sprite?

**Comment by TemporalOroboros** [comment]
@GoldenAlpharex Smoothing is complicated, but the short answer is it needs a new icon state for every junction. This ends up being slightly less than 256 icon states per thing you want to smooth in eight directions. Even using icon colors to cull it down to a base and a greyscale light overlay that's still a lot of icon states.

**Comment by GoldenAlpharex** [comment]
> @GoldenAlpharex Smoothing is complicated, but the short answer is it needs a new icon state for every junction. This ends up being slightly less than 256 icon states per thing you want to smooth in eight directions. Even using icon colors to cull it down to a base and a greyscale light overlay that's still a lot of icon states.

Ah, okay, I wasn't aware it was that tedious.

**Comment by TemporalOroboros** [comment]
It's less tedious and more just the server needs to chuck that data at everyone who logs onto the server so the more of it we have the more expensive it gets.

**Comment by LemonInTheDark** [comment]
ghhhhhhhhhhhhhhhhhhhh

**Comment by maxymax13** [comment]
having these locked away behind stupid reagents/cargo basically means people will never use more them more than once as a hard to get novelty, make the color pickable with tile reskinning
also pacman tiles pacman tiles

**Comment by Farquaar** [comment]
Screenshots?

**Comment by TemporalOroboros** [comment]
@Farquaar 
> Screenshots?

In the dropdown.

**Comment by D4C-420** [comment]
> Alternatively, all simple neon carpet varieties are available at cargo for a small sum.
The arcade cabinet can also dispense one variety of emissive floor as a prize.

Does simple in this case mean does not glow in the dark?

**Comment by TemporalOroboros** [comment]
@D4C-420 
> Does simple in this case mean does not glow in the dark?

Simple just means that the sprites are simple. They still glow. It's to contrast with future designs with more complexity.

**Comment by D4C-420** [comment]
> Simple just means that the sprites are simple. They still glow. It's to contrast with future designs with more complexity.

Okay thanks for clearing that. They look nice by the way.

**Review by Twaticus — APPROVED** [review]

File: code/game/objects/items/stacks/tiles/tile_types.dm [hunk 2]
```diff
@@ -299,6 +299,623 @@
...
+	. += emissive_appearance(neon_icon || icon, neon_icon_state || icon_state, alpha=emissive_alpha)
+
+/obj/item/stack/tile/carpet/neon/worn_overlays(isinhands, icon_file)
+	. = ..()
+	if(!isinhands || !neon_inhand_icon_state)
+		return
+	var/mutable_appearance/neon_overlay = mutable_appearance(icon_file, neon_inhand_icon_state, alpha=src.alpha)
+	neon_overlay.color = neon_color
+	. += neon_overlay
+	// TODO: Figure out how to add an emissive overlay to the inhand.
```
Comment (Mothblocks) [inline code/game/objects/items/stacks/tiles/tile_types.dm:336]: is this TODO now

File: code/game/objects/items/stacks/tiles/tile_types.dm [hunk 3]
```diff
@@ -299,6 +299,623 @@
...
+	desc = "A piece of neon carpet produced under a software error."
+	icon_state = "tile-error"
+	inhand_icon_state = "tile-error"
+	turf_type = /turf/open/floor/fake_error
+	merge_type = /obj/item/stack/tile/fake_error
+	resistance_flags = FLAMMABLE
+
+/obj/item/stack/tile/fake_error/update_overlays()
+	. = ..()
+	. += emissive_appearance(icon, icon_state, alpha=src.alpha)
```
Comment (Mothblocks) [inline code/game/objects/items/stacks/tiles/tile_types.dm:908]:
```suggestion
	. += emissive_appearance(icon, icon_state, alpha = alpha)
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
looks fine

**Comment by optimumtact** [comment]
this ha been on terry for a day, what are your conclusions?

File: code/game/objects/items/stacks/tiles/tile_types.dm (hunk 2 above)
Comment (TemporalOroboros) [reply code/game/objects/items/stacks/tiles/tile_types.dm:336]: I'm still trying to figure that one out. Adding the emissive overlay normally just results in it being squished down to the same plane as the rest of the in hand overlays.

File: code/game/objects/items/stacks/tiles/tile_types.dm (hunk 3 above)
Comment (TemporalOroboros) [reply code/game/objects/items/stacks/tiles/tile_types.dm:908]: Done.

**Comment by LemonInTheDark** [comment]
No noticeable impact
Ignore the headers, they're lies.

This is maptick over time for twice the time this pr's been testmerged
maptick1
This is pop
maptick2

There doesn't seem to be any meaningful variation, so we should be just fine

**Comment by TemporalOroboros** [comment]
There are still a few issues. It looks like some of the emissive blockers are straight up not working. (ex: closets missing the blocker on the door, airlocks missing the blocker on the everything) Additionally, I've been trying to get an emissive overlay on the inhand overlays for the neon carpet tiling and it hasn't been working so far. I'll think of something.

File: code/game/objects/items/stacks/tiles/tile_types.dm (hunk 2 above)
Comment (Mothblocks) [reply code/game/objects/items/stacks/tiles/tile_types.dm]: lmk when you figure it out or give up

**Comment by Kylerace** [comment]
@TemporalOroboros shouldn't the special emissive blockers that mobs get extend to all inhand overlays?

**Comment by LemonInTheDark** [comment]
How are things sitting?

**Comment by TemporalOroboros** [comment]
@LemonInTheDark 
> How are things sitting?

College pre-finals week. I should be able to work on it some more today.

File: code/game/objects/items/stacks/tiles/tile_types.dm (hunk 2 above)
Comment (TemporalOroboros) [reply code/game/objects/items/stacks/tiles/tile_types.dm]: It's done.

**Comment by TemporalOroboros** [comment]
@LemonInTheDark
Changes have been made and the last bits I wanted have been completed.

@Mothblocks 
I figured out how to get the inhands to be emissive. As a nice side effect I have also made moth antenna and lizard frills translucent.

**Pull request merged**